src/main/java/tc/oc/pgm/match/MatchManagerImpl.java

```diff
@@ -139,6 +196,44 @@ private World createMatchWorld(String worldName, MapInfo info, TerrainModule ter
     return world;
   }
 
+  /**
+   * Creates a new world, wich must be done on the primary thread. If already running on the primary
+   * thread, calls createWorld, otherwise, starts a sync task to run create world, and waits until
+   * it's done.
+   *
+   * @param creator The creator for the world
+   * @return A world if it was able to be created, null otherwise
+   */
+  private World createWorld(final WorldCreator creator) {
+    if (server.isPrimaryThread()) return server.createWorld(creator);
+
+    // FIXME: This is pretty dirty, there must be a better way to run & wait on main
+    final Object LOCK = new Object();
+    final AtomicReference<World> world = new AtomicReference<>();
+    final AtomicBoolean isDone = new AtomicBoolean();
+    PGM.get()
+        .getServer()
+        .getScheduler()
+        .runTask(
```

Comment:
This makes me think we need a `Scheduler` for `PGM` (outside `Match` scope) and also have it support async tasks. Because we've been using `getServer#getScheduler` a lot. This is a side note.